### PR TITLE
UCP/WORKER: Add tx/rx wakeup events specification

### LINKED ISSUE #1062: UCX triggers on dummy messages
Two threads on client side: 1st is polling on input messages; 2nd is doing send_tag_nb(…) + progress-loop next to send_tag_nb(…).
After triggering on UCX file descriptor 1st thread makes arm and then progress/probe for messages. 

Piece of polling thread logic:

```
1.  Create UCX Worker()
2.  Arm worker
3.  While (1) {
    a.  epoll_wait(ucx epfd, blocking)
    aa. Arm worker
    b.  do {
        i.  worker_progress()
        ii. msg = probe()
        iii.    if (msg) msg_recv(msg)
    c.  } while (msg);
    d.  goto EPOLL_WAIT
4. }
```

Probe() call often returns NULL, after UCX fd was triggered on event. 

Often means about 700 000 times per second, while about 300 messages of size 1M were sent using UCX during this particular second (using `UCX_TLS=rc UCX_IB_ETH_PAUSE_ON=y UCX_RNDV_THRESH=10000000` parameters).

Setting less granularity of bcopy via `UCX_MAX_BCOPY=500000` fixes this behavior (though is slower).

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
API for #1062 

@shssf @brminich @yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3301/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/832/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2783/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -152,6 +153,7 @@ enum ucp_feature {
 enum ucp_worker_params_field {
     UCP_WORKER_PARAM_FIELD_THREAD_MODE  = UCS_BIT(0), /**< UCP thread mode */
     UCP_WORKER_PARAM_FIELD_CPU_MASK     = UCS_BIT(1), /**< Worker's CPU bitmap */
+    UCP_WORKER_PARAM_FIELD_EVENTS       = UCS_BIT(2), /**< Worker's events bitmap */
```
Comment (brminich) [inline src/ucp/api/ucp.h]: minor: no need for comma

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -577,6 +579,14 @@ typedef struct ucp_worker_params {
      */
     ucs_cpu_set_t           cpu_mask;
 
+    /**
+     * Mask of which events are expected on wakeup.
```
Comment (brminich) [inline src/ucp/api/ucp.h]: minor: I'd swap "which" and "events" words

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -577,6 +579,14 @@ typedef struct ucp_worker_params {
...
+     * This value is optional.
+     * If it's not set (along with its corresponding bit in the field_mask -
+     * UCP_WORKER_PARAM_FIELD_EVENTS), all types of events will be generated on
```
Comment (brminich) [inline src/ucp/api/ucp.h]: maybe "will trigger wakeup"?

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (brminich) [inline src/ucp/api/ucp.h]: minor: no need for comma

File: test/gtest/ucp/test_ucp_wakeup.cc [hunk 4]
```diff
@@ -17,20 +17,119 @@ class test_ucp_wakeup : public ucp_test {
...
+    }
+
+    for (size_t i = msg_count; i < 2*msg_count; ++i) {
+        req[i] = ucp_tag_recv_nb(p_entity[1]->worker(), &recv_data, sizeof(recv_data),
+                                 DATATYPE, TAG, (ucp_tag_t)-1, recv_completion);
+        req_cntr += UCS_PTR_IS_PTR(req[i]) ? 1 : 0;
+    }
+
+    /* wait until all messages are completed */
+    while (req_cntr == comp_cntr) {
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_wakeup.cc]: Maybe you want to check "!=" ?

File: src/ucs/type/wakeup_event.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+#include <ucs/sys/math.h>
+#include <uct/api/uct.h>
+
+typedef enum ucs_wakeup_event_types {
+    UCS_WAKEUP_TX_COMPLETION   = UCS_BIT(0),
+    UCS_WAKEUP_RX_AM           = UCS_BIT(1),
+    UCS_WAKEUP_ALL             = UCS_WAKEUP_TX_COMPLETION | UCS_WAKEUP_RX_AM
+} ucs_wakeup_event_t;
+
+static inline enum uct_wakeup_event_types
```
Comment (brminich) [inline src/ucs/type/wakeup_event.h]: Would move it to UCT, because it uses UCT type which is not desirable in UCS

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/840/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3309/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/841/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2793/ for details (Mellanox internal link).

File: src/ucs/type/wakeup_event.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TYPE_WAKEUP_EVENT_H
+#define UCS_TYPE_WAKEUP_EVENT_H
+
+#include <ucs/sys/math.h>
+
+typedef enum ucs_wakeup_event_types {
```
Comment (yosefe) [inline src/ucs/type/wakeup_event.h]: this should not be in UCS because the concept of "wakeup" is specific for communications - this is not a generic infrastructure

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3338/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/870/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2832/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
Update: the branch is rebased on the master to resolve conflicts

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/962/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2924/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/963/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/979/ for details.

File: src/ucp/api/ucp_def.h [hunk 7]
```diff
@@ -234,5 +235,10 @@ typedef void (*ucp_send_callback_t)(void *request, ucs_status_t status);
 typedef void (*ucp_tag_recv_callback_t)(void *request, ucs_status_t status,
                                         ucp_tag_recv_info_t *info);
 
+typedef enum ucp_wakeup_event_types {
+    UCP_WAKEUP_TX_COMPLETION   = UCS_BIT(0),
+    UCP_WAKEUP_RX_AM           = UCS_BIT(1),
+    UCP_WAKEUP_ALL             = UCP_WAKEUP_TX_COMPLETION | UCP_WAKEUP_RX_AM
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: don't have to create special value for "all"

Comment (yosefe) [inline src/ucp/api/ucp_def.h]: also, pls add documentation the the enum and its fields

File: src/ucp/core/ucp_worker.c [hunk 8]
```diff
@@ -186,9 +186,26 @@ static void ucp_worker_wakeup_context_cleanup(ucp_worker_wakeup_t *wakeup)
...
+
+    if (ucp_events & UCP_WAKEUP_TX_COMPLETION) {
+        uct_events |= UCT_WAKEUP_TX_COMPLETION;
+    }
+
+    if (ucp_events & UCP_WAKEUP_RX_AM) {
+        uct_events |= (UCT_WAKEUP_RX_AM | UCT_WAKEUP_RX_SIGNALED_AM);
+    }
+
+    return (enum uct_wakeup_event_types)uct_events;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: no need for explicit cast

File: src/ucp/core/ucp_worker.c [hunk 9]
```diff
@@ -554,13 +569,16 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
         goto err_destroy_uct_worker;
     }
 
+    events = (params->field_mask & UCP_WORKER_PARAM_FIELD_EVENTS)
+           ? params->events : UCP_WAKEUP_ALL;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: indentation, also ? should be on 1st line

File: test/gtest/ucp/test_ucp_wakeup.cc [hunk 10]
```diff
@@ -17,20 +17,126 @@ class test_ucp_wakeup : public ucp_test {
...
+         * messages arriving, these are not TX-events */
+    }
+
+    if (events & UCP_WAKEUP_RX_AM) {
+        EXPECT_TRUE(polled[1].revents);
+    } else {
+        EXPECT_FALSE(polled[1].revents);
+    }
+
+    for (size_t i = 0; i < 2*msg_count; i++) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wakeup.cc]: could move this to line 97, to avoid need to count completions there

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2943/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_wakeup.cc [hunk 11]
```diff
@@ -105,3 +211,59 @@ UCS_TEST_P(test_ucp_wakeup, signal)
...
 
 UCP_INSTANTIATE_TEST_CASE(test_ucp_wakeup)
+
+class test_ucp_tx_wakeup : public test_ucp_wakeup
+{
+public:
+    static ucp_worker_params_t get_worker_params() {
+        ucp_worker_params_t params = test_ucp_wakeup::get_worker_params();
+        params.field_mask |= UCP_WORKER_PARAM_FIELD_EVENTS;
+        params.events = UCP_WAKEUP_TX_COMPLETION;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wakeup.cc]: could achieve same effect by using generate_test_params_variant(), see for example test_ucp_memheap.cc:19

Comment (evgeny-leksikov) [inline test/gtest/ucp/test_ucp_wakeup.cc]: Not sure, because get_worker_params is a static method (instantiation requirement) but GetParam() should be non-static (gtest API)
Pls correct if I don't understand your idea

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2961/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2962/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2963/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_wakeup.cc (hunk 11 above)
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wakeup.cc]: generate_test_params_variant() is called during gtest startup to create all possible test parameters, it's possible to use it to generate 2 variants for the existing parameters: RX and TX. then use GetParam() during the test to figure on which is it for each test instantiation.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2964/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2966/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2968/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1003/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/999/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/996/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/998/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1001/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/997/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1021/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1022/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1023/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1054/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1059/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3009/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_wakeup.cc [hunk 12]
```diff
@@ -231,9 +231,9 @@ void test_ucp_wakeup_events::do_tx_rx_events_test(unsigned events)
     if (events & UCP_WAKEUP_RX_AM) {
         /* UCP_WAKEUP_RX_AM may cause events on sender side because of internal
          * messages arriving, these are not TX-events */
-        EXPECT_LE(nfd_exp, poll(polled, 2, 200*ucs::test_time_multiplier()));
+        EXPECT_LE(nfd_exp, poll(polled, 2, 10000));
     } else {
-        EXPECT_EQ(nfd_exp, poll(polled, 2, 200*ucs::test_time_multiplier()));
+        EXPECT_EQ(nfd_exp, poll(polled, 2, 10000));
```
Comment (shssf) [inline test/gtest/ucp/test_ucp_wakeup.cc]: I would propose to use `ucs::test_time_multiplier()` because this function make difference between regular and run under valgrind.
For example `1000 * ucs::test_time_multiplier()`

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1065/ for details.

File: test/gtest/ucp/test_ucp_wakeup.cc (hunk 12 above)
Comment (evgeny-leksikov) [inline test/gtest/ucp/test_ucp_wakeup.cc]: I did it by analogy with [this](https://github.com/evgeny-leksikov/ucx/blob/master/test/gtest/uct/test_ib.cc#L445). I think we need to have a single rules for such issues to stabilize the testing

Comment (shssf) [inline test/gtest/ucp/test_ucp_wakeup.cc]: Could you please fix the test, you pointed by the link, in the same way as this test?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3019/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1070/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3022/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1071/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1072/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1073/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3023/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1074/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3024/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.c [hunk 13]
```diff
@@ -554,13 +568,16 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
         goto err_destroy_uct_worker;
     }
 
+    events = (params->field_mask & UCP_WORKER_PARAM_FIELD_EVENTS) ?
+             params->events : (UCP_WAKEUP_TX_COMPLETION | UCP_WAKEUP_RX_AM);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: need also RX_SIGNALED_AM

**Comment by yosefe** [comment]
@evgeny-leksikov pls also squash the comits

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

File: src/ucp/core/ucp_worker.c (hunk 13 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c]: As a separate flag for UCP?
Currently `UCP_WAKEUP_RX_AM` includes `UCT_WAKEUP_RX_SIGNALED_AM`:
```
if (ucp_events & UCP_WAKEUP_RX_AM) {
    uct_events |= (UCT_WAKEUP_RX_AM | UCT_WAKEUP_RX_SIGNALED_AM);
 }
```

Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: right, sorry, pls disregard this comment.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1076/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3026/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1077/ for details.

**Comment by bbenton** [comment]
:+1: for Doc/API

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3027/ for details (Mellanox internal link).

File: test/gtest/uct/test_ib.cc [hunk 14]
```diff
@@ -442,7 +442,7 @@ UCS_TEST_P(test_uct_wakeup_ib, tx_cq)
     send_msg_e1_e2();
 
     /* make sure the file descriptor is signaled once */
-    ASSERT_EQ(poll(&wakeup_fd, 1, 10000), 1);
+    ASSERT_EQ(poll(&wakeup_fd, 1, 1000*ucs::test_time_multiplier()), 1);
```
Comment (shamisp) [inline test/gtest/uct/test_ib.cc]: what is this 1000 magic value

File: src/ucp/api/ucp_def.h [hunk 15]
```diff
@@ -234,5 +234,15 @@ typedef void (*ucp_send_callback_t)(void *request, ucs_status_t status);
...
 
+/**
+ * @ingroup UCP_WORKER
+ * @brief UCP worker wakeup events mask.
+ *
+ * The enumeration allows specifying which events are expected on wakeup
+ */
+typedef enum ucp_wakeup_event_types {
+    UCP_WAKEUP_TX_COMPLETION   = UCS_BIT(0), /**< TX-completions */
+    UCP_WAKEUP_RX_AM           = UCS_BIT(1)  /**< RX-completions */
```
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: This is interface is visible for user ? How TX and RX_AM are mapped to UCP communication directives ?  What TX means for AMOs and GET ? Is it signaled for every fragment ? or once per request...

File: test/gtest/uct/test_ib.cc (hunk 14 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_ib.cc]: We had instabilities last couple weeks in tests which are waiting events on descriptors, so we need to understand which timeout works for any environment including several concurrent jobs on servers. `10000` was added in https://github.com/openucx/ucx/pull/1341, `1000*multiplier` was requested by @shssf and this is a kind of trying to adjust and unify the cases.

File: src/ucp/api/ucp_def.h (hunk 15 above)
Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: Yes, it's visible and optional. It's signaled for every corresponding TX/RX completions the same way as UCT does. Really, there is nothing new, just filtering from UCP level. This is not related directly with UCP requests. See conversation in https://github.com/openucx/ucx/issues/1062

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3030/ for details (Mellanox internal link).

File: src/ucp/api/ucp_def.h (hunk 15 above)
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: @evgeny-leksikov i think @shamisp means that UCP event names and documentation should be consistent with type of operations in UCP level.
for example:
```
UCP_WAKEUP_RMA             = UCS_BIT(0),  /**< Remote memory access send completion */
UCP_WAKEUP_AMO             = UCS_BIT(1),  /**< Atomic operation send completion */
UCP_WAKEUP_TAG_SEND    = UCS_BIT(2), /**< Tag send completion  */
UCP_WAKEUP_TAG_RECV    = UCS_BIT(3),  /**< Tag receive completion */
```

Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: @shamisp @yosefe In scope of this PR I would propose to get rid of transport level specific names in API, then implement signaling by UCP request completions with separate issue and PR.
This way, users will have some events moderation functionality right now and we won't change API later for this need.

Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: @yosefe thanks, but anyway we can't filter type of events with current implementation
(sorry, the page was not updated in my browser :( )

Comment (yosefe) [inline src/ucp/api/ucp_def.h]: we won't filter such events, but only guarantee that they would be delivered.
in the future we could add optimization to filter such events without changing UCP API

Comment (shamisp) [inline src/ucp/api/ucp_def.h]: @yosefe is right. From this PR I cannot understand (without getting into the code and even then it is not obvious) what this event means how it it maps to UCP operations that were invoked a user

Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: So, I'm going to update with the following mapping from UCP to UCT:

```
RMA, AMO, TAG_SEND -> TX_COMPLETION
TAG_RECV -> (RX_AM | RX_SIGNALED_AM)

```
right?

Comment (yosefe) [inline src/ucp/api/ucp_def.h]: yes

Comment (shamisp) [inline src/ucp/api/ucp_def.h]: @yosefe  - it is a bit more complicated. What if an AMO is a combination of GET and another AMO (a.k.a composed protocol). How many events I'm going to get ?

Comment (yosefe) [inline src/ucp/api/ucp_def.h]: then UCP guarantees at least one event would happen, if the user specified AMO flag.

Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: BTW, looks like here possible an issue with rndv-read protocol - let's say, receiver ask only RECV events, so event will be generated only for RTS packet - this is a false event in wrong time

Comment (yosefe) [inline src/ucp/api/ucp_def.h]: right - then i guess would have to signal RNDV manually using the pipe?

Comment (evgeny-leksikov) [inline src/ucp/api/ucp_def.h]: Or implement it with UCP-level file descriptors with corresponding manual signaling on any request completion...

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1083/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1084/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1087/ for details.

File: test/gtest/uct/test_ib.cc (hunk 14 above)
Comment (shamisp) [inline test/gtest/uct/test_ib.cc]: sounds like a hack for hiding a bug ...

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1090/ for details.

File: test/gtest/uct/test_ib.cc (hunk 14 above)
Comment (evgeny-leksikov) [inline test/gtest/uct/test_ib.cc]: I don't think so. I believe that's OK if events are triggered after some time. This is particularly noticeable under valgrind.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1097/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1099/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1107/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1109/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1124/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3059/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe @shamisp pls review again. Now, TAG_RECV includes UCT TX-events as well to handle RNDV case => this PR adds mostly API, but not significant optimization.

**Comment by shamisp** [comment]
Renaming is good but it does not resolve the problem of getting unrelated/dummy events. This is something that we have to discuss.

**Comment by yosefe** [comment]
@shamisp agree, we plan to improve the event filtering in UCT level as well. This PR ends up only adding the API for users.

**Comment by shamisp** [comment]
Let's discuss this on the call.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1141/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3072/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Adding @bbenton (API change)

**Comment by evgeny-leksikov** [comment]
@bbenton please take review updated API

**Comment by bbenton** [comment]
:+1: for API/doc

**Comment by yosefe** [comment]
@bbenton thanks!
@evgeny-leksikov can you please squash to single commit?

**Comment by evgeny-leksikov** [comment]
@yosefe done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1162/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3091/ for details (Mellanox internal link).
